### ISSUE TITLE
Compiling for Windows

### ISSUE THREAD

**Issue by Balgden** [meganz/sdk#196]
Hello,

I'm a Java dev and this is my first time trying to actually compile and all that jazz. My goal is to get the Mega SDK binary for Windows 64bit.
After failing to get MinGW setup on my host machine I decided to fire up a Linux VM instead and follow [this guide](https://github.com/PatrickJBaird/sdk/blob/master/bindings/doc/java/sphinx/source/installsdk.rst) which I found in your open pull requests.

The system is freshly installed from the ubuntu-15.10-desktop-amd64.iso followed by 

```
sudo apt-get update
sudo apt-get dist-upgrade
sudo apt-get install mingw-w64
```

After that I followed the steps from the guide linked above. While I couldn't test the resulting binary, there were no obvious errors. Awesome!

Next, I wanted to cross compile for Win64 with

```
./configure --enable-java --host=x86_64-w64-mingw32 --with-java-include-dir=/usr/lib/jvm/java-7-openjdk-amd64/include/
```

which just ends in 

```
configure: error: cryptopp/cryptlib.h header not found or not usable
```

libcrypto++-dev is installed as evident by the fact that the linux compile worked fine.

Could you maybe tell me what I'm doing wrong/missing?

Here are the config.logs from the successful linux build as well as from the unsuccesful cross compile attempt.
[configlogs.zip](https://github.com/meganz/sdk/files/67432/configlogs.zip)

**Comment by wizzard** [comment]
Hello,

you need to point configure script to where all MinGW dependencies libraries and headers are located (by default it checks fo libraries /usr/lib and /usr/include).

**Comment by Balgden** [comment]
Thank you very much. Although MinGW automatically searches in the correct locations, your comment made me realize that installing the dependencies on Linux won't help me with cross-compiling for Windows. I had to go and cross-compile all dependencies as well.

Configure runs through now, however there are still compile errors when running make.
The first one is in include/mega/win32/megasys.h, where you include Wincrypt.h and Winhttp.h. Note the uppercase W in the filename. That causes MinGW not to find the files in question even though they are present. You might want to correct that unless there are reasons to keep the includes the way they are.

Once that was fixed, I ran into more compile errors, which apparently have already been reported (here)[https://github.com/meganz/sdk/issues/201]. I will try again once a solution for that issue has been found.

**Comment by Balgden** [comment]
Thanks to wizzard for finding out that --without-sodium works around the compile errors from https://github.com/meganz/sdk/issues/201.

Building still doesn't run through cleanly though.
First, there are a bruch of warnings like this for what looks like every file:
```
CXX      src/src_libmega_la-megaapi_impl.lo
src/megaapi_impl.cpp:1:0: warning: -fPIC ignored for target (all code is position independent)
 ﻿/**
 ^
  CXX      src/src_libmega_la-megaapi.lo
src/megaapi.cpp:1:0: warning: -fPIC ignored for target (all code is position independent)
 /**
 ^
  CXX      src/win32/src_libmega_la-fs.lo
src/win32/fs.cpp:1:0: warning: -fPIC ignored for target (all code is position independent)
```
From what I understood these are probably safe to ignore and just annoying.

At the end I get this though:
```
CXXLD    src/libmega.la
libtool: link: warning: undefined symbols not allowed in x86_64-w64-mingw32 shared libraries
x86_64-w64-mingw32-ar: `u' modifier ignored since `D' is the default (see `U')
/usr/bin/swig2.0 -c++ -java -package nz.mega.sdk -outdir ./bindings/java/nz/mega/sdk -I./include -o bindings/java/megaapi_wrap.cpp bindings/megaapi.i
  CXX      bindings/java/bindings_java_libmegajava_la-megaapi_wrap.lo
bindings/java/megaapi_wrap.cpp:1:0: warning: -fPIC ignored for target (all code is position independent)
 /* ----------------------------------------------------------------------------
 ^
  CXXLD    bindings/java/libmegajava.la

*** Warning: This system can not link to static lib archive ./src/libmega.la.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have.
libtool: link: warning: undefined symbols not allowed in x86_64-w64-mingw32 shared libraries
x86_64-w64-mingw32-ar: `u' modifier ignored since `D' is the default (see `U')
```
As already suggested by these warnings, no shared file (libmegajava.so) is created, only a libmegajava.a. From looking at megaJNI.java it looks like libmegajava.so is exaclty the file I need though.

My current configure line is:
    ./configure --enable-java --host=x86_64-w64-mingw32 --enable-shared --prefix=/home/adler/megasdk --disable-examples --without-freeimage --with-java-include-dir=/usr/lib/jvm/java-8-oracle/include/ --with-cryptopp=/home/adler/dependencies/cryptopp --with-zlib=/home/adler/dependencies/zlib --with-sqlite=/home/adler/dependencies/sqlite --with-winhttp=/home/adler/dependencies/winhttp

Any ideas would be appreciated.

**Comment by pohutukawa** [comment]
@Balgden, please be aware, though, that soon libsodium will be a prerequisite for using the SDK. So, even though it may not be necessary now, it might be a good idea to figure out how to build with it already.

@wizzard: We need to suss this out for our different Jenkins builds, too.

**Comment by Balgden** [comment]
@pohutukawa: According to what javiserrano wrote (here)[https://github.com/meganz/sdk/issues/201#issuecomment-168609250] the build with libsodium should succeed once the chat keys branch has been merged into master. So, until that time comes and it still doesn't work, I'll assume that I used libsodium correctly.
You're of course right. The build should work with all options enabled. For the sake of completeness I'll probably try to get freeimage and the examples to build as well some time.
